### ISSUE TITLE
DataReceived never goes off & more

### ISSUE THREAD

**Issue by lukeacat** [jchristn/SimpleTcp#48]
I'm making a simple chatapp with TCP.

`Connected` was fired perfectly, and I could send my JSON.

I was finished with the backend, and I added some stuff to DataReceived, but it didn't fire even though I know I'm sending json.

Also, I sent a bunch of disconnects, then connnects. They after a second of doing this stacked up and became from `{hello:info}` to this `{hello:info}{hello:info}{hello:info}`

Thanks for maintaining!,
Luke.

**Comment by jchristn** [comment]
Hi @lukeacat ill be back at my computer tonight. At first glance you may want to use WatsonTcp instead since it provides framing. Will respond in more detail later￼

**Comment by jchristn** [comment]
Hi @lukeacat sorry for the delay.  

The reason you may be seeing multiple messages is that SimpleTcp provides no framing layer.  What this means is that your client might be sending a 12 byte message (i.e. ```{hello:info}```) but it's just going into a big memory buffer on the receiver side (server) and the server has no idea how many bytes to read from the socket.  SimpleTcp just has a background loop that continues tries to read as much as it can (up to the amount specified in ```Settings.StreamBufferSize```, in bytes), and when it gets some amount of data back from the socket, it fires ```DataReceived```.

Framing (see https://blog.stephencleary.com/2009/04/message-framing.html) encapsulates your data with a smaller header that indicates how many bytes of data are being sent.  In your case, framing your ```{hello:info}``` message could be as simple as ```12:{hello:info}```, and the server could read one byte at a time until it hits the delimiter (i.e. ```:```), and parses what it has read into an ```int```, and then it knows to read 12 bytes, i.e. ```{hello:info}```.

So, what is likely happening is the TCP layer is receiving data from the server and making it available for read.  When SimpleTcp receives that data, it drains as much as it can (up to, again, ```Settings.StreamBufferSize```) and fires the event.

If you need message-by-message events, you need framing - there's no way around that.  That's why I would recommend looking at WatsonTcp.

Cheers, Joel

**Comment by lukeacat** [comment]
This doesn't address the `DataReceived` issue. That's what I was mostly concerned about.

I already made a working fix for framing on the server side. 

Thanks, Luke!

**Comment by jchristn** [comment]
DataReceived will only fire when there is data actually read from the socket.  Is there any way you could provide some sample code that reproduces this?

Also the framing on WatsonTcp is pretty straightforward.  I could walk you through how to implement framing to support it if you've already done it client side.

**Comment by lukeacat** [comment]
Hmm, I can put my game on Github (It's in its early stages). I'll report back, since the project should be opensource.

**Comment by lukeacat** [comment]
https://github.com/lukeacat/SimpleGame Here's the project!

Thanks for maintaining as always, Joel!

**Comment by lukeacat** [comment]
Even more Info: 
When using the provided node.js client (test.js) inside the backend folder with a name, everything works how I want it to work.

**Comment by lukeacat** [comment]
Another clue: I tried this with Websockets (I recoded the backend and tried websockets)
and the same thing is happening.. WTF?

**Comment by lukeacat** [comment]
(talking about websockets)
![image](https://user-images.githubusercontent.com/35845382/108708129-b23e6d80-7519-11eb-9ed2-12f2c6a3a132.png)
Mm, I am so confused. I started the server in Administrator mode right? Yeah, I did that, but the c# client didn't work. Ok, well that's fine. Now, I made a quick web browser client, and it WORKED! It doesn't work with normal permissions.. What's going on? It's not a antivirus, I assure you, I've stopped all of my firewalls / antiviruses from doing anything. 
![image](https://user-images.githubusercontent.com/35845382/108708608-53c5bf00-751a-11eb-8aa0-0a54afd662a3.png)

**Comment by jchristn** [comment]
Hi @lukeacat in https://github.com/lukeacat/SimpleGame/blob/master/frontend/Form1.cs can you try using ```await client.SendAsync```?

Am I correct in assuming that you're only using SimpleTcp on the client side?  The only thing I could think of is that more than one message is making it into the receive buffer in between attempts to read from the socket.  That would cause data to be bundled together in the ```DataReceived``` event.  

Firing the ```DataReceived``` event can cause a (very, very minor) delay between socket reads, and if two messages arrive in the socket between these reads, data can be commingled.  Another reason why framing is really important (side note: websockets do provide framing natively).

I'm having a really hard time reproducing the issue even when running on separate nodes, but I'm also using the SimpleTcp client and server (not the JS backend you have in your repo).

Is there a way in your backend to minimize the runtime of ```socket.on('data', function(data) {```?  I imagine that with JS being single threaded, anything introducing a delay in the completion of this event will allow data to queue up in the receive buffer from multiple client sends.  It looks like the code 1) parses JSON to object and then 2) on "connect", broadcasts a message to everyone (which may add to the runtime of the method)

Thoughts?  Cheers, Joel

**Comment by lukeacat** [comment]
Hey Joel, I'm closing this in light on my recent discoveries:

` What went wrong?:`
Earlier, I got CS0120 on the DataReceived part (theForm was made in a different thread).
This meant that dataRecieved was started. 

`How'd you fix it?:`
I added `await` to `client.sendAsync`, and that's all I really did.
I don't know what went wrong last time, but now it works.

Also, to debug I added MessageBox.Show().

`Does it work now?:`
Well, we still don't know what happened, but everything else seems fine for now.

`Answering those questions:`
`1.` Yes, I'm only using STCP on Client Side.
`2 & 3:` Desync is possible, but it shouldn't drop entire messages. Should it?
`4.` Minimizing runtime in JS isn't really possible, it's as good as you write as good as you get.

Thanks for your continued help Joel,
luke out!

**Comment by jchristn** [comment]
Hi @lukeacat wow, well, at very least I am really glad it's working for you now.

I can't speak to why messages got dropped, that indicates that they didn't arrive, or, an exception is being encountered.  You can try setting ```client.Logger``` to output to a messagebox, console, etc, and see if something shows up there that might indicate if there is an exception.  I'm betting that ```await``` was keeping the thread waiting on the completion of ```SendAsync```, and what was likely happening before was that ```void Connected``` was exiting prior to the unawaited ```SendAsync``` completing.  I'm doubtful that ```void Connected``` completing prior to unawaited ```SendAsync``` would cause the task to be garbage collected, but I could be wrong (not an expert in forms/WPF and threading).

Cheers Luke and have a good one!
